staging/src/k8s.io/client-go/tools/cache/fifo.go

```diff
@@ -275,48 +319,58 @@ func (f *FIFO) IsClosed() bool {
 func (f *FIFO) Pop(process PopProcessFunc) (interface{}, error) {
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	for {
-		for len(f.queue) == 0 {
-			// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
-			// When Close() is called, the f.closed is set and the condition is broadcasted.
-			// Which causes this loop to continue and return from the Pop().
-			if f.closed {
-				return nil, ErrFIFOClosed
-			}
-
-			f.cond.Wait()
-		}
-		id := f.queue[0]
-		f.queue = f.queue[1:]
-		if f.initialPopulationCount > 0 {
-			f.initialPopulationCount--
-		}
-		item, ok := f.items[id]
-		if !ok {
-			// Item may have been deleted subsequently.
-			continue
-		}
-		delete(f.items, id)
-		err := process(item)
-		if e, ok := err.(ErrRequeue); ok {
-			f.addIfNotPresent(id, item)
-			err = e.Err
+
+	for f.queue.Len() == 0 {
+		// When the queue is empty, invocation of Pop() is blocked until new item is enqueued.
+		// When Close() is called, the f.closed is set and the condition is broadcasted.
+		// Which causes this loop to continue and return from the Pop().
+		if f.closed {
+			return nil, ErrFIFOClosed
 		}
-		return item, err
+
+		f.cond.Wait()
```

Comment:
so it is waiting / blocking until new item is enqueued while holding the lock f.lock.Lock(), where there be deadlock?

Reply:
`It won't deadlock. Here are reasons: \
When we create a FIFO, we use function NewFIFO as below:
```
func NewFIFO(keyFunc KeyFunc) *FIFO {
	f := &FIFO{
		items:   map[string]interface{}{},
		queue:   list.New(),
		keyFunc: keyFunc,
	}
	f.cond.L = &f.lock
	return f
}
```
and `f.cond.L = &f.lock`. it means the Lock in f.cond is f.lock. Thus, when we use `Wait` in `Pop` function as below:

```
func (f *FIFO) Pop(process PopProcessFunc) (interface{}, error) {
	f.lock.Lock()
	defer f.lock.Unlock()

	for f.queue.Len() == 0 {
		...
		f.cond.Wait()
	}
}
```
`Wait` will atomically unlocks c.L(f.lock) and suspends execution of the calling goroutine.
Therefore, there will be no deadlock.